src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1291,19 +1305,45 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
                 _readLen = 0;
             }
 
-            int n = _bufferSize - _writePos;
-            if (numBytes <= n)
+            int remainingBuffer = _bufferSize - _writePos;
+
+            // if the incoming write fits in the remaining buffer, treat
+            // as synchronous since memcopy is so fast
+            if (numBytes <= remainingBuffer)
             {
                 if (_writePos == 0) _buffer = new byte[_bufferSize];
                 Buffer.BlockCopy(array, offset, _buffer, _writePos, numBytes);
                 _writePos += numBytes;
-                
+
                 return Task.CompletedTask;
             }
 
-            if (_writePos > 0)
-                FlushWrite(false);
+            // if the incoming write fits in the current buffer + next buffer,
+            // only perform a single write to flush the existing buffer and fill
+            // the next buffer with the remainder
+            if (numBytes <= (_bufferSize + remainingBuffer))
+            {
+                if (_writePos == 0) _buffer = new byte[_bufferSize];
+
+                Buffer.BlockCopy(array, offset, _buffer, _writePos, remainingBuffer);
+
+                byte[] oldBuffer = _buffer;
+                _buffer = new byte[_bufferSize];
+
+                _writePos = (numBytes - remainingBuffer);
+                Buffer.BlockCopy(array, offset + remainingBuffer, _buffer, 0, _writePos);
+
+                return WriteInternalCoreAsync(oldBuffer, 0, _bufferSize, cancellationToken);
+            }
+
+            // If there's data in the existing buffer, and the incoming write would overflow it and
+            // the subsequent buffer, chain the flush operation with the subsequent write
+            else if(_writePos > 0)
+                return FlushWriteAsync(cancellationToken)
+                    .ContinueWith((_) => WriteInternalCoreAsync(array, offset, numBytes, cancellationToken));
```

Comment:
(That said, there are streams in the framework that say you're not allowed to reuse the same buffer you provided until the async operation has completed... including the base Stream class, so maybe FileStream does the same thing?)